src/ir/module-utils.cpp

```diff
@@ -537,44 +543,53 @@ InsertOrderedMap<HeapType, size_t> getHeapTypeCounts(Module& wasm,
 
     // We've found all the types there are to find without considering more
     // control flow types. Consider one more control flow type and repeat.
-    for (; controlFlowIt != counts.controlFlowSignatures.end();
-         ++controlFlowIt) {
+    for (; controlFlowIt != info.controlFlowSignatures.end(); ++controlFlowIt) {
       auto& [sig, count] = *controlFlowIt;
       if (auto it = seenSigs.find(sig); it != seenSigs.end()) {
-        counts.counts[it->second] += count;
+        info.info[it->second].useCount += count;
       } else {
         // We've never seen this signature before, so add a type for it.
         HeapType type(sig);
         noteNewType(type);
-        counts.counts[type] += count;
+        info.info[type].useCount += count;
         break;
       }
     }
   }
 
-  return counts.counts;
-}
-
-void setIndices(IndexedHeapTypes& indexedTypes) {
-  for (Index i = 0; i < indexedTypes.types.size(); i++) {
-    indexedTypes.indices[indexedTypes.types[i]] = i;
+  if (classify == ClassifyAction::Classify) {
+    classifyTypes(wasm, info.info);
   }
+
+  return std::move(info.info);
 }
 
-InsertOrderedSet<HeapType> getPublicTypeSet(Module& wasm) {
-  InsertOrderedSet<HeapType> publicTypes;
+namespace {
+
+void classifyTypes(Module& wasm,
+                   InsertOrderedMap<HeapType, HeapTypeInfo>& types) {
+  // We will need to traverse the types used by public types and mark them
+  // public as well.
+  std::vector<HeapType> workList;
 
   auto notePublic = [&](HeapType type) {
     if (type.isBasic()) {
-      return;
+      return false;
     }
     // All the rec group members are public as well.
+    bool inserted = false;
     for (auto member : type.getRecGroup()) {
-      if (!publicTypes.insert(member)) {
-        // We've already inserted this rec group.
-        break;
+      if (auto it = types.find(member); it != types.end()) {
+        if (it->second.classification == TypeClassification::Public) {
+          // We've already inserted this rec group.
```

Comment:
I find these two lines hard to follow. Why does the classification being public show that we've already inserted the rec group? Perhaps the comment should be more like "we've already inserted this rec group, and marked it as public"?